File: aitest_kit/agent/protocol.py
"""Versioned JSONL protocol shared by the Python control plane and Pi Worker."""
from __future__ import annotations

import json
import re
import uuid
from dataclasses import dataclass
from typing import Any, Mapping


PROTOCOL_VERSION = 1
SENSITIVE_KEYS = {
    "api_key",
    "apikey",
    "authorization",
    "cookie",
    "password",
    "secret",
    "token",
}
_SECRET_VALUE_PATTERNS = (
    re.compile(r"(?i)\bBearer\s+[^\s\"']+"),
    re.compile(r"\bsk-[A-Za-z0-9_-]{8,}"),
)


class ProtocolError(RuntimeError):
    """Raised when a JSONL message violates the AITest protocol."""

    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code


def redact(value: Any) -> Any:
    """Return a JSON-safe copy with credential-shaped values removed."""
    if isinstance(value, Mapping):
        redacted: dict[str, Any] = {}
        for raw_key, child in value.items():
            key = str(raw_key)
            normalized = key.lower().replace("-", "_")
            redacted[key] = "[REDACTED]" if normalized in SENSITIVE_KEYS else redact(child)
        return redacted
    if isinstance(value, list):
        return [redact(child) for child in value]
    if isinstance(value, tuple):
        return [redact(child) for child in value]
    if isinstance(value, str):
        rendered = value
        for pattern in _SECRET_VALUE_PATTERNS:
            rendered = pattern.sub("[REDACTED]", rendered)
        return rendered
    return value


@dataclass(frozen=True)
class ProtocolMessage:
    protocol_version: int
    id: str
    type: str
    payload: dict[str, Any]

    @classmethod
    def create(
        cls,
        message_type: str,
        payload: Mapping[str, Any] | None = None,
        *,
        message_id: str | None = None,
    ) -> "ProtocolMessage":
        if not message_type:
            raise ProtocolError("INVALID_ENVELOPE", "message type must be a non-empty string")
        return cls(
            protocol_version=PROTOCOL_VERSION,
            id=message_id or str(uuid.uuid4()),
            type=message_type,
            payload=dict(payload or {}),
        )
    @classmethod
    def from_line(cls, line: str) -> "ProtocolMessage":
        try:
            raw = json.loads(line)
        except (TypeError, json.JSONDecodeError) as exc:
            raise ProtocolError("INVALID_JSON", "worker output is not valid JSON") from exc
        if not isinstance(raw, dict):
            raise ProtocolError("INVALID_ENVELOPE", "protocol message must be a JSON object")
        version = raw.get("protocol_version")
        if version != PROTOCOL_VERSION:
            raise ProtocolError(
                "UNSUPPORTED_PROTOCOL_VERSION",
                f"unsupported protocol version: {version!r}",
            )
        message_id = raw.get("id")
        message_type = raw.get("type")
        payload = raw.get("payload")
        if (
            not isinstance(message_id, str)
            or not message_id
            or not isinstance(message_type, str)
            or not message_type
            or not isinstance(payload, dict)
        ):
            raise ProtocolError("INVALID_ENVELOPE", "protocol envelope fields are invalid")
        return cls(version, message_id, message_type, payload)

    def to_line(self) -> str:
        return json.dumps(
            {
                "protocol_version": self.protocol_version,
                "id": self.id,
                "type": self.type,
                "payload": redact(self.payload),
            },
            ensure_ascii=False,
            separators=(",", ":"),
        )

File: aitest_kit/agent/runtime_seed/pi_worker/src/protocol.ts
export const PROTOCOL_VERSION = 1;

const COMMAND_TYPES = new Set([
  "initialize",
  "prompt",
  "permission_decision",
  "abort",
  "shutdown",
]);

const SENSITIVE_KEYS = new Set([
  "api_key",
  "apikey",
  "authorization",
  "cookie",
  "password",
  "secret",
  "token",
]);
const SECRET_VALUES = new Set<string>();

export interface ProtocolMessage {
  protocol_version: number;
  id: string;
  type: string;
  payload: Record<string, unknown>;
}

export class ProtocolFailure extends Error {
  code: string;

  constructor(code: string, message: string) {
    super(message);
    this.name = "ProtocolFailure";
    this.code = code;
  }
}

export function parseCommand(line: string): ProtocolMessage {
  let value: unknown;
  try {
    value = JSON.parse(line);
  } catch {
    throw new ProtocolFailure("INVALID_JSON", "input line is not valid JSON");
  }
  if (!isRecord(value)) {
    throw new ProtocolFailure("INVALID_ENVELOPE", "protocol message must be a JSON object");
  }
  if (value.protocol_version !== PROTOCOL_VERSION) {
    throw new ProtocolFailure(
      "UNSUPPORTED_PROTOCOL_VERSION",
      `unsupported protocol version: ${String(value.protocol_version)}`,
    );
  }
  if (typeof value.id !== "string" || value.id.length === 0) {
    throw new ProtocolFailure("INVALID_ENVELOPE", "message id must be a non-empty string");
  }
  if (typeof value.type !== "string" || value.type.length === 0) {
    throw new ProtocolFailure("INVALID_ENVELOPE", "message type must be a non-empty string");
  }
  if (!COMMAND_TYPES.has(value.type)) {
    throw new ProtocolFailure("UNKNOWN_MESSAGE_TYPE", `unknown message type: ${value.type}`);
  }
  if (!isRecord(value.payload)) {
    throw new ProtocolFailure("INVALID_ENVELOPE", "message payload must be a JSON object");
  }
  return value as unknown as ProtocolMessage;
}

export function createMessage(
  id: string,
  type: string,
  payload: Record<string, unknown> = {},
): ProtocolMessage {
  return { protocol_version: PROTOCOL_VERSION, id, type, payload };
}

export function serializeMessage(message: ProtocolMessage): string {
  return JSON.stringify(redact(message));
}

export function registerSecret(value: string | undefined): void {
  if (value && value.length >= 4) SECRET_VALUES.add(value);
}

export function redact(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map((item) => redact(item));
  }
  if (isRecord(value)) {
    const result: Record<string, unknown> = {};
    for (const [key, child] of Object.entries(value)) {
      const normalized = key.toLowerCase().replaceAll("-", "_");
      result[key] = SENSITIVE_KEYS.has(normalized) ? "[REDACTED]" : redact(child);
    }
    return result;
  }
  if (typeof value === "string") {
    let rendered = value;
    for (const secret of SECRET_VALUES) {
      rendered = rendered.replaceAll(secret, "[REDACTED]");
    }
    return rendered
      .replace(/\bBearer\s+[^\s"']+/gi, "[REDACTED]")
      .replace(/\bsk-[A-Za-z0-9_-]{8,}/g, "[REDACTED]");
  }
  return value;
}

export function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}
